src/wasm/wasm-type.cpp

```diff
@@ -103,38 +166,75 @@ void Type::init(const std::vector<Type>& types) {
     *this = types[0];
     return;
   }
+  id = canonicalize(TypeDef(tuple));
+}
 
-  // Add a new type if it hasn't been added concurrently
-  std::lock_guard<std::mutex> lock(mutex);
-  auto indexIt = indices.find(types);
-  if (indexIt != indices.end()) {
-    id = indexIt->second;
-  } else {
-    auto vec = std::make_unique<std::vector<Type>>(types);
-    id = uintptr_t(vec.get());
-    constructedTypes.push_back(std::move(vec));
-    assert(id > _last_value_type);
-    indices[types] = id;
+Type::Type(const Signature& signature) {
+  id = canonicalize(TypeDef(signature));
+}
+
+Type::Type(const Struct& struct_) {
+#ifndef NDEBUG
+  for (Field f : struct_.fields) {
+    assert(f.type.isSingle() && f.type.isConcrete());
   }
+#endif
+  id = canonicalize(TypeDef(struct_));
 }
 
-Type::Type(std::initializer_list<Type> types) { init(types); }
+Type::Type(const Array& array) {
+#ifndef NDEBUG
+  assert(array.element.type.isSingle() && array.element.type.isConcrete());
+#endif
+  id = canonicalize(TypeDef(array));
+}
 
-Type::Type(const std::vector<Type>& types) { init(types); }
+bool Type::isTuple() const {
+  if (id > _last_value_type) {
+    std::lock_guard<std::mutex> lock(mutex);
+    auto it = complexLookup.find(id);
+    if (it != complexLookup.end()) {
+      return it->second.getKind() == TypeDef::TupleKind;
+    }
+  }
+  return false;
+}
+
+bool Type::isRef() const {
+  if (id > _last_value_type) {
+    std::lock_guard<std::mutex> lock(mutex);
+    auto it = complexLookup.find(id);
+    if (it != complexLookup.end()) {
+      switch (it->second.getKind()) {
+        case TypeDef::SignatureKind:
+        case TypeDef::StructKind:
+        case TypeDef::ArrayKind:
+          return true;
+        default:
+          return false;
+      }
+    }
+  }
+  return id >= funcref && id <= exnref;
+}
 
 size_t Type::size() const { return expand().size(); }
 
-const std::vector<Type>& Type::expand() const {
-  if (id <= _last_value_type) {
-    return basicTypes[id];
-  } else {
-    return *(std::vector<Type>*)id;
+const TypeList& Type::expand() const {
+  std::lock_guard<std::mutex> lock(mutex);
+  auto it = complexLookup.find(id);
+  if (it != complexLookup.end()) {
+    auto& typeDef = it->second;
+    if (typeDef.getKind() == TypeDef::TupleKind) {
+      return typeDef.tupleDef.tuple.types;
+    }
   }
+  WASM_UNREACHABLE("invalid type");
```

Comment:
This used to be more efficient before when expanding basic types because we had a separate array of `basicTypes` tuples to access without a mutex, but I'm not sure how common expanding basic types actually is, given that anything else needs a mutex anyway to determine whether the complex type is a tuple. Any preference?

Reply:
Oh, OF COURSE, just understood the idea. Made it cast to `TypeDef*` now.